Add locality load balancing docs to traffic management concepts (#3593)
istio/istio.io: content/docs/concepts/traffic-management/index.md

check. In such an event, the service instance will be immediately removed
from the caller's load balancing pool.

## Locality Load Balancing

{{< tip >}}

> This feature is currently experimental and can be enabled by setting the `PILOT_ENABLE_LOCALITY_LOAD_BALANCING` environment variable in all Pilot instances.

Author: Literally anything apart from an empty string.

Currently, locality is automatically populated by the service discovery platform. In Kubernetes, a pod's locality is determined via the [well-known labels for region and zone](https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/#failure-domain-beta-kubernetes-io-region) on the node it is deployed. Sub-zone is not a concept that exists in Kubernetes and as a result is not configured.

### Locality Prioritized Load Balancing

The default behavior of Locality Load Balancing is Locality Prioritized Load Balancing. In this mode, Istio will tell Envoy to prioritize traffic to endpoints that most closely match the locality of the Envoy forwarding the request. When all endpoints are healthy, this means that requests will remain within the same locality. When endpoints become unhealthy, traffic will spillover to endpoints in the next prioritized locality and so on until all localities are receiving traffic. Details of the exact percentages can be found [in the Envoy documentation](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/load_balancing/priority#priority-levels).

> A typical prioritization for an Envoy with a locality of `us-west/zone2` would be as follows:

Reviewer: ```suggestion
A typical prioritization for an Envoy with a locality of `us-west/zone2` is as follows:
```

- Priority 0: `us-west/zone2`
- Priority 1: `us-west/zone1`, `us-west/zone3`
- Priority 2: `us-east/zone1`, `us-east/zone/2`, `eu-west/zone1`

> Prioritization is hierarchical, in that it matches for sub-zone, zone and region in that order; endpoints in the same zone but different regions are not considered local to one another.

Reviewer: ```suggestion
The hierarchy of prioritization matches in the following order:

 1. Sub-zone
 1. Zone
 1. Region
 
Services in the same zone but different regions are not considered local to one another.
```

> #### Overriding Locality Failover

Reviewer: ```suggestion
#### Overriding the Locality Fail-over
```

> In some cases users may need to constrain traffic failover so that failing over to any endpoint globally when there are not enough healthy endpoints in the same region does not apply. This is useful when failing over traffic across regions would not improve service health or may need to be restricted for other reasons like regulatory controls. This can be achieved by configuring the mesh `LocalityLoadBalancerSetting.Failover` configuration detailed [here](/docs/reference/config/istio.mesh.v1alpha1/#LocalityLoadBalancerSetting-Failover).

Reviewer: ```suggestion
Sometimes, you need to constrain the traffic fail-over to avoid sending traffic to
a service across the globe when there are not enough healthy endpoints in the
same region. This behavior is useful when sending fail-over traffic across regions
would not improve service health or many other reasons including regulatory controls.
To constrain traffic to a region, use the mesh `LocalityLoadBalancerSetting.Failover`
configuration detailed in the [Locality load balancing reference guide](/docs/reference/config/istio.mesh.v1alpha1/#LocalityLoadBalancerSetting-Failover).
```

check. In such an event, the service instance will be immediately removed
from the caller's load balancing pool.

## Locality Load Balancing

{{< warning >}}
This feature is currently experimental. To enabled it, set the `PILOT_ENABLE_LOCALITY_LOAD_BALANCING` environment variable in all Pilot instances.

> {{< /warning >}}

Reviewer 2: This is not the right place for this kind of information. The concepts doc doesn't just talk about Pilot. It talks about what Istio can do in general, agnostic of platforms.. So, to add locality load balancing here, you simply add a subsection and talk about how users get out of box locality load balancing based on geographic regions, etc. (omit all flags stuff/experimental stuff/etc.). 

Then in a task (locality based routing), you specify that people need to enable it explicitly for 1.1 and that they should annotate the nodes, or add locality to their service entries. You don't need the usual verify if it works, blah.. 

cc @frankbu

Reviewer 3: Another alternative is to move this into the operations guide - add an entry `Locality Load Balacancing` with this content [here](https://preliminary.istio.io/help/ops/traffic-management/). Then in the `Load Balancing` concepts section just add a single sentence or two with a reference up to the operations guide.
